egs/wsj/s5/steps/nnet3/fvector/allocate_examples.py

```diff
@@ -0,0 +1,234 @@
+             "than a third of the maximum length {1} from the utt2len file ".format(
+            args.max_frames_per_chunk, max_length))
+
+# create the map form ori-utt-id to all kinds of utt-id. The ori-utt-id is the 
+# index, which is same with the elements in utt_ids[]
+f = open(args.oriutt2allutt, "r");
+if f is None:
+    sys.exit("Error opening oriutt2allutt file " + str(args.oriutt2allutt));
+utt_map = {}
+for line in f:
+    a = line.split()
+    if len(a) < 3:
+        sys.exit("bad line in oriutt2allutt file " + line);
+    tmp_list = []
+    for i in range(1, len(a)):
+        tmp_list.append(a[i])
+    tuple_list = tuple(tmp_list)
+    utt_map[a[0]]=tuple_list
+f.close()
+
+    
+# this function returns a random integer utterance index, limited to utterances
+# above a minimum length in frames, with probability proportional to its length.
+def RandomUttAtLeastThisLong(min_length):
+    while True:
+        i = random.randrange(0, num_utts)
+        # read the next line as 'with probability lengths[i] / max_length'.
+        # this allows us to draw utterances with probability with
+        # prob proportional to their length.
+        if lengths[i] > min_length and random.random() < lengths[i] / float(max_length):
+            return i
+
+
+# given an utterance length utt_length (in frames) and two desired chunk lengths
+# (length1 and length2) whose sum is <= utt_length,
+# this function randomly picks the starting points of the chunks for you.
+# the chunks may appear randomly in either order.
+def GetRandomOffsets(utt_length, length):
+    if length > utt_length:
+        sys.exit("code error: tot-length > utt-length")
+    free_length = utt_length - length
+    offset = random.randrange(0, free_length + 1)
+    return offset
+
+
+# this function randomly choose two utt-id form utt_map depending on ori-utt-id
+def ChoosePairs(ori_utt_id):
+    this_tuple = utt_map[ori_utt_id]
+    while True:
+        first_index = random.randint(0, len(this_tuple) - 1)
+        second_index = random.randint(0, len(this_tuple) - 1)
+        if first_index != second_index:
+            break
+    utt_a = this_tuple[first_index]
+    utt_b = this_tuple[second_index]
+    return (utt_a, utt_b)
+
+
+# each element of all_egs (one per archive) is
+# an array of 2-tuples (utterance-index, offset)
+all_egs= []
+
+prefix = ""
+if args.prefix != "":
+  prefix = args.prefix + "_"
+
+for archive_index in range(args.num_archives):
+    tot_length = 2 * args.frames_per_chunk
+    this_num_egs = (args.frames_per_iter / tot_length) + 1
+    this_egs = [ ] # this will be an array of 2-tuples (utterance-index, start-frame).
+    for n in range(this_num_egs):
+        utt_index = RandomUttAtLeastThisLong(args.frames_per_chunk)
+        utt_len = lengths[utt_index]
+        offset = GetRandomOffsets(utt_len, args.frames_per_chunk)
+        this_egs.append( (utt_index, offset) )
+    all_egs.append(this_egs)
```

Comment:
From line 180 to 189:
I generate a 'egs list' for each archive. Each item in the 'egs list' is a tuple (utt, offset). 
In later code, I will use each tuple to generate the final information which will be parsed by nnet3-fvector-get-egs.cc. 
I will use 'utt' to randomly select two utterances from '<ori-utt-id> <p1-utt-id> ... <pn-utt-id>' to generate '<source-utterance1> <source-utterance2>'. And 'offset' is used to assign "<offset>", which means the distance between the beginning of utterance and the start frame.
So, at last, I will generate the information as followed:
<source-utterance1> <source-utterance2> <relative-archive-index> <absolute-archive-index> <offset> <frame-length>

I'm not sure the running speed. I just use a few utterance to test the code is right or not. I use the same method with David to do this part.